addons/mail/controllers/discuss/gif.py

```diff
@@ -0,0 +1,58 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import json
+import requests
+from odoo.http import request, route, Controller
+
+
+class DiscussGifController(Controller):
+
+    def _request_gifs(self, url):
+        try:
+            response = requests.get(url, timeout=3)
+        except requests.exceptions.RequestException:
+            return False
+        if response.status_code != requests.codes.ok:
+            return False
+        return response
+
+    @route("/discuss/gif/search", type="json", auth="public", cors="*")
+    def search(self, search_term, locale="en", country="US", position=False):
+        config = request.env["ir.config_parameter"].sudo().tenor_query_params()
+        url = f"https://tenor.googleapis.com/v2/search?q={search_term}&key={config.get('api_key')}&client_key={config.get('client_key')}&limit={config.get('tenor_gif_limit')}&contentfilter={config.get('content_filter')}&locale={locale}&country={country}&media_filter=tinygif"
+        if position:
+            url = f"{url}&pos={position}"
+        response = self._request_gifs(url)
+        if response:
+            return json.loads(response.content)
+
+    @route("/discuss/gif/categories", type="json", auth="public", cors="*")
+    def categories(self, locale="en", country="US"):
+        config = request.env["ir.config_parameter"].sudo().tenor_query_params()
+        url = f"https://tenor.googleapis.com/v2/categories?key={config.get('api_key')}&client_key={config.get('client_key')}&contentfilter={config.get('content_filter')}&locale={locale}&country={country}"
+        response = self._request_gifs(url)
+        if response:
+            return json.loads(response.content)
+
+    @route("/discuss/gif/add_favorite", type="json", auth="user")
+    def add_favorite(self, tenor_gif_id):
+        request.env["discuss.gif.favorite"].create({"tenor_gif_id": tenor_gif_id})
+
+    def _gif_posts(self, ids):
+        config = request.env["ir.config_parameter"].sudo().tenor_query_params()
+        url = f"https://tenor.googleapis.com/v2/posts?key={config.get('api_key')}&client_key={config.get('client_key')}&ids={','.join(ids)}&media_filter=tinygif"
+        response = self._request_gifs(url)
+        if response:
+            return json.loads(response.content)["results"]
+
+    @route("/discuss/gif/favorites", type="json", auth="user")
+    def get_favorites(self, offset=0):
+        tenor_gif_ids = request.env["discuss.gif.favorite"].search([("create_uid", "=", request.env.user.id)], limit=20, offset=offset)
+        return {
+            "results": self._gif_posts(tenor_gif_ids.mapped("tenor_gif_id")) or [],
+            "offset": offset + 20
```

Comment:
I don't understand the point of this result if it's a static value depending on user input `offset` only, it could be computed in JS directly?

Reply:
Because I completly forgot to implement the load more for favorite. Yes this can be done in Javascript :)